pkg/apis/ceph.rook.io/v1/types.go

```diff
@@ -221,6 +221,28 @@ type ClusterSpec struct {
 	// +optional
 	// +nullable
 	LogCollector LogCollectorSpec `json:"logCollector,omitempty"`
+
+	// CSI Driver Options applied per cluster.
+	CsiDriverOptions CSIDriverOptionsSpec `json:"csiDriverOptions,omitempty"`
+}
+
+// CSIDriverOptionsSpec defines CSI Driver settings applied per cluster.
+type CSIDriverOptionsSpec struct {
```

Comment:
Is there a struct defined in the CSI repo that we could use in the CRD type? If Rook uses the CSI type directly, we don't need to translate between the Rook and CSI types. When we read from the CR, it can be immediately serialized to the configmap exactly as the csi driver expects it, and no translation code would be needed. Perhaps that won't be possible, but would like to see if it's an option.

Reply:
> Is there a struct defined in the CSI repo that we could use in the CRD type? If Rook uses the CSI type directly, we don't need to translate between the Rook and CSI types. When we read from the CR, it can be immediately serialized to the configmap exactly as the csi driver expects it, and no translation code would be needed. Perhaps that won't be possible, but would like to see if it's an option.

Currently the struct is not exported. https://github.com/ceph/ceph-csi/blob/593d9c3b605adc742940a012736184662315983d/internal/util/csiconfig.go#L40.

This needs to be done at cephcsi first and changes will need to be done at a lot of places in rook repo.
I'll open a separate follow up pr to import structs from cephcsi repo and use it in rook.
Even with the struch, some translation would definitely be required.
That change will be internal only, we don't need to block this pr.

Reply:
I guess we need different structs anyway since there are differences where Rook will generate the mons, and we obviously don't want those specified in the cluster CR under the csi settings.

Reply:
I think it would be challenging to use the whole CSI config struct directly in the CephCluster. I imagine that many of the fields don't make sense for users to modify since Rook will fill them in dynamically based on the Ceph cluster config and status.

If there are any specific data types that CSI uses that have special validation, it could be useful for CSI to provide those as an API that Rook can use to pre-validate configs before they reach CSI. I've noticed that a lot of users struggle with CSI debugging when they have mis-configured something because they only look at Rook operator logs and events rather than CSI logs and events, and I think that would be helpful.

And also, that can certainly be something that gets small improvements over time.